Test the install procedure

dentarg opened the pull request:
Should catch things like #40.

dentarg committed:
Test the install procedure

Should catch things like #40.

dentarg commented:
Added `--user` because I got:

```
$ make deps
pip install -r test_requirements.txt
Collecting nose (from -r test_requirements.txt (line 1))
  Downloading https://files.pythonhosted.org/packages/99/4f/13fb671119e65c4dce97c60e67d3fd9e6f7f809f2b307e2611f4701205cb/nose-1.3.7-py2-none-any.whl (154kB)
    100% |████████████████████████████████| 163kB 12.5MB/s
Collecting betamax (from -r test_requirements.txt (line 2))
  Downloading https://files.pythonhosted.org/packages/d1/54/820de0a06d938917103f7d91c024403b9ae521d74e1191fe6f91a70089c9/betamax-0.8.1-py2.py3-none-any.whl
Requirement already satisfied: python-dateutil in /Users/dentarg/Library/Python/2.7/lib/python/site-packages (from -r test_requirements.txt (line 3)) (2.7.5)
Requirement already satisfied: requests>=2.0 in /usr/local/lib/python2.7/site-packages (from betamax->-r test_requirements.txt (line 2)) (2.20.1)
Requirement already satisfied: six>=1.5 in /usr/local/lib/python2.7/site-packages (from python-dateutil->-r test_requirements.txt (line 3)) (1.10.0)
Requirement already satisfied: urllib3<1.25,>=1.21.1 in /usr/local/lib/python2.7/site-packages (from requests>=2.0->betamax->-r test_requirements.txt (line 2)) (1.23)
Requirement already satisfied: chardet<3.1.0,>=3.0.2 in /Users/dentarg/Library/Python/2.7/lib/python/site-packages (from requests>=2.0->betamax->-r test_requirements.txt (line 2)) (3.0.4)
Requirement already satisfied: idna<2.8,>=2.5 in /usr/local/lib/python2.7/site-packages (from requests>=2.0->betamax->-r test_requirements.txt (line 2)) (2.7)
Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python2.7/site-packages (from requests>=2.0->betamax->-r test_requirements.txt (line 2)) (2018.8.24)
Installing collected packages: nose, betamax
Could not install packages due to an EnvironmentError: [Errno 13] Permission denied: '/usr/local/man'
Consider using the `--user` option or check the permissions.

make: *** [deps] Error 1
```

dentarg commented:
However, even after adding `--user`, it didn't work. Not sure why, but after `pip freeze | xargs pip uninstall -y`, it did work. I did not try without user (as we are using `--user` elsewhere and I think it is needed).

dentarg commented:
Hmm... where did Travis go?

Here's the build anyway: https://travis-ci.org/twingly/twingly-search-api-python/builds/499326142

dentarg commented:
Error for both 2.7 and 3.5:

> Can not perform a '--user' install. User site-packages are not visible in this virtualenv.

walro commented on Makefile:
```diff
@@ -3,7 +3,8 @@ help:
 deps:
-	pip install -r test_requirements.txt
+	pip install --user --requirement test_requirements.txt
+	pip install .
```
hm, is this desirable? It will install the "repo version" of `twingly-search-api-python` in addition to the dependencies.

dentarg replied:
Ah, you mean you do not want it installed on your local system, makes some sense.

I think it is desirable on CI, as it uses `setup.py` then.

walro replied:
I think `make deps` should install just deps, nothing else. Even on Travis.

dentarg replied:
But `pip install .` is also installing dependencies :)

walro replied:
Hehe, a bit of a stretch. Anyway, seems like we agree that installing in local system is unexpected either way.

dentarg replied:
I would like to use the same setup locally and on Travis to be able to catch errors like #40 and I want it to be the same so it doesn't go out of sync, but that seems impossible if we don't want to do `pip install .` locally. We should add something that can run the whole testing locally with Docker then.

dentarg replied:
The whole point of this PR was to do `pip install .` instead of `pip install -r requirements.txt`, so I don't see going back to `pip install -r requirements.txt` on Travis as an option. Do you have any other thoughts on it @walro?

walro replied:
I see your point, meanwhile I would move this line back to `.travis.yml` as it will at least make travis pick up on the errors. 

Locally I would just add a line to the readme under development: `pip install -r requirements.txt` for now.

I find the risk for it to go out of date pretty low as locally you will find out soon enough that you are lacking required packages.

walro replied:
I didn't see your comment before writing mine, but I think I answered it anyway:)

dentarg replied:
Sure, let's do it like that.

I think I will remove the Makefile altogether, as we really can't share any stuff with it between CI and local environment. We also don't use Makefiles in our other projects, so our know-how with them is low (one idea I had was to branch on [`TRAVIS=true`](https://docs.travis-ci.com/user/environment-variables/#default-environment-variables) somewhere, but I don't know exactly what to write without reading some docs).

dentarg replied:
OK, changed my mind, I will keep the Makefile, but make two separate "deps" targets (one for CI, one for local env).

walro replied:
sounds good :)

dentarg commented:
I wondering if it is worth supporting running the tests in macOS on our workstations... it would be more similar to CI if we ran them in a one off Docker container.

dentarg committed:
Avoid installing the "repo" version locally

https://github.com/twingly/twingly-search-api-python/pull/42#discussion_r260802495

dentarg committed:
Test with all supported Python versions

https://devguide.python.org/devcycle/#end-of-life-branches

dentarg committed:
CI: Python 3.7 is only available in Ubuntu 16.04

https://github.com/travis-ci/travis-ci/issues/9831
https://github.com/travis-ci/travis-ci/issues/9815

dentarg commented:
All is green at https://travis-ci.org/twingly/twingly-search-api-python/builds

Seems like Travis having some issues today
